project_6

Egor1253 opened the pull request:
Прошу проверить работу после pull request.

Egor1253 committed:
Добрый день.
Помогите разобраться пожалуйста с кодом!!!

Commit:
Добрый день.
Помогите разобраться пожалуйста с кодом!!!

Commit:
Добрый вечер!
Прошу посмотреть изменения в коде.

Commit:
Добрый вечер!
Прошу посмотреть изменения в коде, от 08.12.23

Commit:
Добрый вечер!
Прошу посмотреть изменения в коде, от 08.12.23

mkouznetsov-git commented on src/test/java/CatTest.java:
```diff
@@ -0,0 +1,35 @@
+@RunWith(MockitoJUnitRunner.class)
+public class CatTest {
+    @Mock
+    Cat cat;
```
Нет смысла мокировать тестируемый объект

Egor1253 replied:
Понял, убрал мок.

mkouznetsov-git commented on src/test/java/FelineTest.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Test
+    public void herbivore() throws Exception {
+        feline.getFood("Травоядное");
```
Что делает эта строчка и для чего она здесь нужна?

Egor1253 replied:
Убрал лишнюю строчку.

mkouznetsov-git commented on src/test/java/FelineTest.java:
```diff
@@ -0,0 +1,44 @@
+    @Test
+    public void herbivore() throws Exception {
+        feline.getFood("Травоядное");
+        when(feline.eatMeat()).thenReturn(of("Трава", "Различные растения"));
```
Зачем проверять для кошачьих, что они едят траву и растения?

Egor1253 replied:
Не знал куда впихнуть эту проверку, написано все должно быть покрыто тестами.

mkouznetsov-git replied:
Эта часть касается класса Animal, можешь покрыть тестами его тоже. Но в задании не требуется стопроцентного покрытия тестами этого класса. Главное -- покрыть полностью классы Feline, Cat и Lion.

mkouznetsov-git commented on src/test/java/FelineTest.java:
```diff
@@ -0,0 +1,44 @@
+    @Test
+    public void felineGetFamily() {
+        feline.getFamily();
+        Mockito.verify(feline).getFamily();
```
Непонятно, что проверяет этот тест

Egor1253 replied:
Он проверяет работоспособность метода.

mkouznetsov-git commented on src/test/java/FelineTest.java:
```diff
@@ -0,0 +1,44 @@
+        Mockito.verify(feline).getFamily();
+    }
+    @Test
+    public void felineGetKittens(){
```
Сначала ты указываешь возвращать единицу при вызове метода, потом вызываешь метод и проверяешь, что он возвращает единицу. В чём смысл теста?

Egor1253 replied:
Если проверяемая часть кода еще не работает, то я принудительно показываю ей - верни такое - то значение, затем проверяю, что это значение вернулось. Все как в теории, там проверка ради проверки, а до меня не доходит как написать проверку от себя, потому как только учусь и первый раз пишу код, вот и копирую, изгаляюсь...)

mkouznetsov-git replied:
Ничего страшного. Главное -- учиться. Я задаю вопросы не для того, чтобы покритиковать. Я задаю специальные вопросы для того, чтобы ты сам понял, как и какие вопросы нужно себе самому задавать, когда будешь писать тесты в будущем.
Здесь по факту получается, что ты тестируешь библиотеку мокирования. Это бессмысленно.
Вообще, не нужно мокировать тестируемый объект, иначе ты подменяешь его реальную реализацию (которую тебе надо проверить) логикой мока.

mkouznetsov-git commented on src/test/java/LionTest.java:
```diff
@@ -0,0 +1,39 @@
+    @Mock
+    Lion lionSpy = Mockito.spy(Lion.class);
+
+    // Проверка - есть ли у Льва львята?
```
Это не проверка, есть ли у льва львята. Это проверка, что мы вызвали метод getKittens 3 раза.

Egor1253 replied:
Поправил.

Egor1253 replied:
Комментарий остался от того, что код в этом месте уже раз 20 переписывал.

mkouznetsov-git replied:
Данная проверка не имеет смысла. Обычно используют такой подход, когда подразумевают, что при вызове метода, который подвергается тестированию, будет далее вызван другой метод (вызовы которого мы проверяем через spy). Но тут мы напрямую вызываем метод и затем проверяем, что мы его вызвали 3 раза. Мы и так его только что вызвали 3 раза -- чего тут проверять? )

mkouznetsov-git commented on src/test/java/LionTest.java:
```diff
@@ -0,0 +1,39 @@
+    }
+    // проверил неизвестный вид животного.
+    @Test
+    public void whoIsIt() throws Exception {
```
В этом тесте нет проверок

mkouznetsov-git commented on src/test/java/LionTest.java:
```diff
@@ -0,0 +1,39 @@
+
+    // проверил, хищник ли лев.
+    @Test
+    public void whoIsLion() throws Exception {
```
В этом тесте нет проверок

Commit:
Максим, пожалуйста подскажите если есть существенные ошибки в коде, я уже окончательно запутался в реализации.

mkouznetsov-git commented on src/test/java/LionTest.java:
```diff
@@ -0,0 +1,50 @@
+    // проверил неизвестный вид животного.
+    @Test
+    public void whoIsIt() throws Exception {
+        doThrow(new IOException("Неизвестный вид животного, используйте значение Травоядное или Хищник"))
```
Мы в этом классе проверяем класс Lion. Не нужно его заставлять делать что-то другое. Нам надо проверить его настоящее поведение, как оно сейчас запрограммировано.

mkouznetsov-git commented on src/test/java/CatTest.java:
```diff
@@ -0,0 +1,33 @@
+
+    // Проверяю: Хищник ли кот?
+    @Test
+    public void whoIsCat() throws Exception {
```
Тест падает с ошибкой:
org.mockito.exceptions.base.MockitoException: Unable to create mock instance of type 'Cat'

mkouznetsov-git commented on src/test/java/CatTest.java:
```diff
@@ -0,0 +1,33 @@
+
+    // проверяю: что говорит кот.
+    @Test
+    public void catSays() {
```
Тест падает с ошибкой:
org.mockito.exceptions.base.MockitoException: Unable to create mock instance of type 'Cat'

mkouznetsov-git commented on src/test/java/LionTest.java:
```diff
@@ -0,0 +1,50 @@
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class LionTest {
```
В этом классе все тесты падают с ошибками.

Commit:
Поправлено по замечаниям.

mkouznetsov-git commented on src/main/java/com/example/Lion.java:
```diff
@@ -2,11 +2,12 @@
     boolean hasMane;
-
-    public Lion(String sex) throws Exception {
+    public Lion () {}
```
Для чего эта строчка?

Commit:
1) При отсутствии конструктора без параметров: public Lion() {}, падают тесты в классе:LionTest.
2) У меня лион в параметризованном тесте еще есть, какую еще проверку написать, не совсем понимаю, вроде все покрыто тестами, написал в параметризованный тест ошибку, для вызова я else (огурец), идеи закончились.

Commit:
Поправил в соответствии с замечаниями.
